### PR TITLE
internal/dag: Add support for ingressroute delegation

### LINKED ISSUE #448: ingressroute: add support for delegation
Add support for delegated ingress route objects.

### PR DESCRIPTION
**Pull request by alexbrand**
Fixes #448 

```
IngressRoute (IR) resources can either point to backend services, or delegate
a path to another IR. While building the DAG out of the IRs and routes
within each, we determine whether the route delegates to another IR.
When it does, we find the referenced IR and process it accordingly.

This processing is applied recursively until we traverse all the edges
and find a leaf route, or a route that does not delegate to another IR.
This is the base case, in which we grab the services, and add them to
the routes of the corresponding VHost.

Signed-off-by: Alexander Brand <alexbrand09@gmail.com>
```

To Do:
- [x] More tests
- [x] Code/approach review

### CODE REVIEW

**Comment by davecheney** [comment]
LGTM. You'll have to rebase to pick up the change in #460. Thanks

File: internal/dag/dag.go [hunk 1]
```diff
@@ -349,17 +375,24 @@ func (d *DAG) recompute() *dag {
...
 		}
-	}
 
-	_d := new(dag)
-	for _, vh := range _vhosts {
-		_d.roots = append(_d.roots, vh)
-	}
-	for _, svh := range _svhosts {
-		_d.roots = append(_d.roots, svh)
+		// otherwise, if the route is delegating to another ingressroute, find it and process it.
```
Comment (stevesloka) [inline internal/dag/dag.go]: Could there be an error condition that a route has a delegate and a service defined? I think @rosskukulinski came up with this the other day. Given that, we should throw it out as an error.

File: internal/dag/dag.go [hunk 2]
```diff
@@ -349,17 +375,24 @@ func (d *DAG) recompute() *dag {
...
+		if route.Delegate.Name != "" {
+			namespace := route.Delegate.Namespace
+			if namespace == "" {
+				// we are delegating to another IngressRoute in the same namespace
+				namespace = ir.Namespace
+			}
+			dir, ok := d.ingressroutes[meta{name: route.Delegate.Name, namespace: namespace}]
```
Comment (stevesloka) [inline internal/dag/dag.go]: Probably for the next PR but if not found then should be an error status condition.

Comment (alexbrand) [reply internal/dag/dag.go]: Yep. I think we can handle this in the next PR.

File: internal/dag/dag.go (hunk 1 above)
Comment (stevesloka) [reply internal/dag/dag.go]: Per slack discussion, we'll handle this in status pr to 1. check for this status and 2. set it on the IngressRoute object

**Pull request merged**